### ISSUE TITLE
[HOLD for payment 2024-11-11] [$250]  Invoice - In the first invoice created, the text " Expense " is not shown

### ISSUE THREAD

**Issue by IuliiaHerets** [Expensify/App#48465]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.28
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**Issue reported by:** Applause Internal Team
 

## Action Performed:
1. Launch app
2. Tap fab -- new workspace
3. Navigate to LHN
4. Tap fab -- Send invoice
5. Enter an amount and tap next
6. Select a user
7.Tap send invoice
8. Via plus icon -- send invoice, create 2 more invoice
9. Note preview of all invoice created

## Expected Result:
In the first invoice created, the text " Expense " is not shown

## Actual Result:
In the first invoice created, the text " Expense ut it is displayed in other invoice created.

## Workaround:
Unknown

## Platforms:

- [x] Android: Native
- [x] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

https://github.com/user-attachments/assets/cb33e405-030e-4125-a913-64c4601dbad3

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021837147636277169515</li>
        <li>Upwork Job ID: 1837147636277169515</li>
        <li>Last Price Increase: 2024-10-18</li>
<li>Automatic offers: </li>
<ul>
<li>aimane-chnaif | Reviewer | 104521419</li>
<li>nkdengineer | Contributor | 104521423</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @puneetlath</details>

**Comment by IuliiaHerets** [comment]
@puneetlath  FYI I haven't added the External label as I wasn't 100% sure about this issue. Please take a look and add the label if you agree it's a bug and can be handled by external contributors

**Comment by nkdengineer** [comment]
The same RCA with https://github.com/Expensify/App/issues/48238

**Comment by puneetlath** [comment]
Thanks @nkdengineer. Going to put on hold for that issue.

**Comment by puneetlath** [comment]
Looks like it doesn't need to be anymore. I removed the hold and I'll ask QA to re-test.

**Comment by m-natarajan** [comment]
Able to reproduce

https://github.com/user-attachments/assets/9f10ebe3-01db-4e2f-b440-e2c3e30993ed

**Comment by nkdengineer** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
In the first invoice created, the text " Expense ut it is displayed in other invoice created.

### What is the root cause of that problem?
1. The first send invoice is created with `merchant` as `(none)` then it works as expected. After that, the transaction draft still exists because this logic

https://github.com/Expensify/App/blob/b3675dac1b8e003088a56a2bd7ebc8e504581cbd/src/pages/iou/request/IOURequestStartPage.tsx#L63-L69

2. When we start the second send invoice, we clear the `transactionDraft` first and then open the `IOURequestStartPage` which will initiate the money request. That problem starts here, the draft transaction is updated after we call `initMoneyRequest` (let's see the log below, the transaction change log is logged [here](https://github.com/Expensify/App/blob/b3675dac1b8e003088a56a2bd7ebc8e504581cbd/src/libs/actions/IOU.ts#L189)). 


https://github.com/Expensify/App/blob/b3675dac1b8e003088a56a2bd7ebc8e504581cbd/src/libs/actions/IOU.ts#L382-L385


https://github.com/user-attachments/assets/1f91d69a-58a1-4c9d-ad18-2f82b746eb09


We move to this case because the transaction still exists as mentioned in point 1. Then the transaction draft is init with `merchant` as `undefined` because the current `transaction` value in the storage is `null` then we send `merchant` as an empty string to `SendInvoice` API. As expected, BE returns the `merchant` as `Expense`.


https://github.com/Expensify/App/blob/b3675dac1b8e003088a56a2bd7ebc8e504581cbd/src/libs/actions/IOU.ts#L323

### What changes do you think we should make in order to solve the problem?
We have some options to solve this issue

1. We can add `Promise.all` to ensure the draft transaction is cleared before we start another money request

```
function startMoneyRequest(iouType: ValueOf<typeof CONST.IOU.TYPE>, reportID: string, requestType?: IOURequestType, skipConfirmation = false) {
    Promise.all([clearMoneyRequest(CONST.IOU.OPTIMISTIC_TRANSACTION_ID, skipConfirmation)]).then(() => {
        switch (requestType) {
            case CONST.IOU.REQUEST_TYPE.MANUAL:
                Navigation.navigate(ROUTES.MONEY_REQUEST_CREATE_TAB_MANUAL.getRoute(CONST.IOU.ACTION.CREATE, iouType, CONST.IOU.OPTIMISTIC_TRANSACTION_ID, reportID));
                return;
            case CONST.IOU.REQUEST_TYPE.SCAN:
                Navigation.navigate(ROUTES.MONEY_REQUEST_CREATE_TAB_SCAN.getRoute(CONST.IOU.ACTION.CREATE, iouType, CONST.IOU.OPTIMISTIC_TRANSACTION_ID, reportID));
                return;
            case CONST.IOU.REQUEST_TYPE.DISTANCE:
                Navigation.navigate(ROUTES.MONEY_REQUEST_CREATE_TAB_DISTANCE.getRoute(CONST.IOU.ACTION.CREATE, iouType, CONST.IOU.OPTIMISTIC_TRANSACTION_ID, reportID));
                return;
            default:
                Navigation.navigate(ROUTES.MONEY_REQUEST_CREATE.getRoute(CONST.IOU.ACTION.CREATE, iouType, CONST.IOU.OPTIMISTIC_TRANSACTION_ID, reportID));
        }
    });
}
```

https://github.com/Expensify/App/blob/b3675dac1b8e003088a56a2bd7ebc8e504581cbd/src/libs/actions/IOU.ts#L382-L385

2. Pass the `transaction` value from the component to `initMoneyRequest` instead of getting it from `allTransactionDrafts`

https://github.com/Expensify/App/blob/b3675dac1b8e003088a56a2bd7ebc8e504581cbd/src/pages/iou/request/IOURequestStartPage.tsx#L63-L69

3. Fallback the merchant here to `CONST.TRANSACTION.PARTIAL_TRANSACTION_MERCHANT` instead of empty string

https://github.com/Expensify/App/blob/b3675dac1b8e003088a56a2bd7ebc8e504581cbd/src/libs/actions/IOU.ts#L1862

4. Add another check to not display the merchant if it's the default merchant as we do [here](https://github.com/Expensify/App/blob/b3675dac1b8e003088a56a2bd7ebc8e504581cbd/src/components/ReportActionItem/MoneyRequestPreview/MoneyRequestPreviewContent.tsx#L158)

```
if (formattedMerchant && formattedMerchant !== CONST.TRANSACTION.DEFAULT_MERCHANT && formattedMerchant !== CONST.TRANSACTION.PARTIAL_TRANSACTION_MERCHANT && formattedMerchant !== CONST.TRANSACTION.DEFAULT_MERCHANT) { 

```
https://github.com/Expensify/App/blob/b3675dac1b8e003088a56a2bd7ebc8e504581cbd/src/components/ReportActionItem/ReportPreview.tsx#L383
### What alternative solutions did you explore? (Optional)

5. We can add a `useEffect` with a cleanup function in `IOURequestStartPage` to clear the transaction draft when this page is unmounted if the `action` param is `create`

**Comment by nkdengineer** [comment]
@aimane-chnaif I think the expected here is the second invoice shouldn't show expense because we don't type the merchant.

**Comment by nkdengineer** [comment]
@aimane-chnaif  Please reset the cache and reproduce with the step in the OP, you can see the first send invoice has an empty merchant but the second send invoice has the merchant as `Expense`. You can see my RCA for more detail.

**Comment by nkdengineer** [comment]
@aimane-chnaif The expected should be all sent invoice should not show `Expense`.

**Comment by nkdengineer** [comment]
Updated the [problem](https://github.com/Expensify/App/issues/48465#issuecomment-2372019035).

**Comment by puneetlath** [comment]
I agree with that. I don't think they should be showing "expense".

**Comment by puneetlath** [comment]
@aimane-chnaif what's next here?

**Comment by nkdengineer** [comment]
@aimane-chnaif I will provide a test branch soon.

**Comment by nkdengineer** [comment]
The test branch is [here](https://github.com/nkdengineer/App/tree/fix/48465).

**Comment by nkdengineer** [comment]
@aimane-chnaif I don't think it's a regression from my branch. I can reproduce this on the latest main.

<img width="1512" alt="Screenshot 2024-10-17 at 10 02 33" src="https://github.com/user-attachments/assets/50912476-ab31-42ef-b397-034b16e26980">

**Comment by nkdengineer** [comment]
> you meant

Yes.

> And is there any reliable repro step?


Here is the step:

1. Make sure the `transactionsDraft_1` is cleared when you send the first invoice by resetting cache and restart or logout and login again or you can manually clear this key in OnyxDB
2. After that when you create the send invoice the first time you will never see `Expense`.

<img width="1512" alt="Screenshot 2024-10-17 at 12 52 32" src="https://github.com/user-attachments/assets/8de17fd0-5bf4-4617-9551-91e6f8b813c5">

**Comment by nkdengineer** [comment]
@aimane-chnaif You mean you always see `Expense` when creating the send invoice on the latest main.

**Comment by nkdengineer** [comment]
> @nkdengineer bug is still reproducible when manually Onyx.merge('transactionsDraft_1', {merchant: null}) on confirmation screen.

@aimane-chnaif As I mentioned in the proposal, if we send invoice with `merchant` param is empty string, BE will return merchant as `Expense`.

In the case above, because you clear the merchant then it's fallback to empty here

https://github.com/Expensify/App/blob/b3675dac1b8e003088a56a2bd7ebc8e504581cbd/src/libs/actions/IOU.ts#L1862

When we init the money request, merchant is fallback as `(none)` and it will be hidden in the preview.
https://github.com/Expensify/App/blob/b3675dac1b8e003088a56a2bd7ebc8e504581cbd/src/libs/actions/IOU.ts#L360

**Comment by nkdengineer** [comment]
> Fallback the merchant here to CONST.TRANSACTION.PARTIAL_TRANSACTION_MERCHANT instead of empty string

> Add another check to not display the merchant if it's the default merchant as we do [here](https://github.com/Expensify/App/blob/b3675dac1b8e003088a56a2bd7ebc8e504581cbd/src/components/ReportActionItem/MoneyRequestPreview/MoneyRequestPreviewContent.tsx#L158)

I also mentioned some another ways to fix this bug

**Comment by nkdengineer** [comment]
@aimane-chnaif I updated the test branch.

**Comment by nkdengineer** [comment]
The code change in the proposal for this solution is wrong with this idea so I updated it a bit. We need to remove this case and return empty string if `numberOfRequests` is 1 to prevent `iou.expenseCount` is shown with `numberOfRequests` is `1`

https://github.com/Expensify/App/blob/b3675dac1b8e003088a56a2bd7ebc8e504581cbd/src/components/ReportActionItem/ReportPreview.tsx#L389-L391

**Comment by nkdengineer** [comment]
@aimane-chnaif Yes that will not cause the previous issue because the description is set in this case.

**Comment by nkdengineer** [comment]
@aimane-chnaif I think the current behavior on the main branch is also a bug because you can see if we submit a normal expense with the merchant as `Expense`, this merchant doesn't show in the money request preview.

https://github.com/user-attachments/assets/2f7ff738-6815-4555-90c0-ea01e1c1d9c2

**Comment by puneetlath** [comment]
That seems fine to me.

**Comment by nkdengineer** [comment]
@aimane-chnaif this [PR](https://github.com/Expensify/App/pull/51217) is ready for review

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by puneetlath** [comment]
All paid. Thanks y'all!
